File: apps/api-server/app/schemas/transaction.py
"""
Dot-Store V2.1 收支记录数据模式
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Dict
from pydantic import BaseModel, Field, ConfigDict


class TransactionCreate(BaseModel):
    """
    收支记录创建模式
    """
    type: str = Field(..., pattern="^(income|expense)$", description="类型：income或expense")
    category: str = Field(..., min_length=1, max_length=64, description="分类名称")
    amount: Decimal = Field(..., gt=0, description="金额，必须大于0")
    order_id: Optional[int] = Field(None, description="关联订单ID")
    note: Optional[str] = Field(None, max_length=500, description="备注")
    attachment_url: Optional[str] = Field(None, max_length=256, description="凭证图片URL")


class TransactionUpdate(BaseModel):
    """
    收支记录更新模式
    """
    category: Optional[str] = Field(None, min_length=1, max_length=64, description="分类名称")
    amount: Optional[Decimal] = Field(None, gt=0, description="金额")
    order_id: Optional[int] = Field(None, description="关联订单ID")
    note: Optional[str] = Field(None, max_length=500, description="备注")
    attachment_url: Optional[str] = Field(None, max_length=256, description="凭证图片URL")


class TransactionResponse(BaseModel):
    """
    收支记录响应模式
    """
    model_config = ConfigDict(from_attributes=True)
    
    id: int = Field(..., description="收支记录ID")
    user_id: int = Field(..., description="用户ID")
    type: str = Field(..., description="类型：income或expense")
    category: str = Field(..., description="分类名称")
    amount: Decimal = Field(..., description="金额")
    order_id: Optional[int] = Field(None, description="关联订单ID")
    note: Optional[str] = Field(None, description="备注")
    attachment_url: Optional[str] = Field(None, description="凭证图片URL")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    created_by: int = Field(..., description="创建人ID")


class TransactionListResponse(BaseModel):
    """
    收支记录列表响应模式
    """
    items: List[TransactionResponse] = Field(..., description="收支记录列表")
    total: int = Field(..., description="总数")
    page: int = Field(1, description="当前页码")
    page_size: int = Field(10, description="每页数量")


class TransactionFilters(BaseModel):
    """
    收支记录筛选条件
    """
    start_date: Optional[datetime] = Field(None, description="开始日期")
    end_date: Optional[datetime] = Field(None, description="结束日期")
    type: Optional[str] = Field(None, pattern="^(income|expense)$", description="类型筛选")
    category: Optional[str] = Field(None, description="分类筛选")
    page: int = Field(1, ge=1, description="页码")
    page_size: int = Field(10, ge=1, le=100, description="每页数量")


class TransactionSummary(BaseModel):
    """
    收支汇总统计模式
    """
    income: Decimal = Field(0, description="总收入")
    expense: Decimal = Field(0, description="总支出")
    net_profit: Decimal = Field(0, description="净利润")
    categories: Dict[str, Decimal] = Field(default_factory=dict, description="分类统计")


class TransactionCategoryCreate(BaseModel):
    """
    收支分类创建模式
    """
    name: str = Field(..., min_length=1, max_length=64, description="分类名称")
    type: str = Field(..., pattern="^(income|expense)$", description="分类类型：income或expense")
    description: Optional[str] = Field(None, max_length=256, description="分类描述")


class TransactionCategoryUpdate(BaseModel):
    """
    收支分类更新模式
    """
    name: Optional[str] = Field(None, min_length=1, max_length=64, description="分类名称")
    description: Optional[str] = Field(None, max_length=256, description="分类描述")


class TransactionCategoryResponse(BaseModel):
    """
    收支分类响应模式
    """
    model_config = ConfigDict(from_attributes=True)
    
    id: int = Field(..., description="分类ID")
    user_id: int = Field(..., description="用户ID")
    name: str = Field(..., description="分类名称")
    type: str = Field(..., description="分类类型")
    description: Optional[str] = Field(None, description="分类描述")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")


class UploadResponse(BaseModel):
    """
    文件上传响应模式
    """
    url: str = Field(..., description="文件URL")
    filename: str = Field(..., description="文件名")

File: apps/frontend/src/types/transaction.ts
/**
 * 收支记录相关类型定义
 */

/**
 * 收支记录接口
 */
export interface Transaction {
  id: number;
  user_id: number;
  type: 'income' | 'expense';
  category: string;
  amount: string;
  order_id: number | null;
  note: string | null;
  attachment_url: string | null;
  created_at: string;
  updated_at: string;
  created_by: number;
}

/**
 * 收支分类接口
 */
export interface TransactionCategory {
  id: number;
  user_id: number;
  name: string;
  type: 'income' | 'expense';
  description: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * 创建收支记录参数
 */
export interface TransactionCreateParams {
  type: 'income' | 'expense';
  category: string;
  amount: number | string;
  order_id?: number;
  note?: string;
  attachment_url?: string;
}

/**
 * 更新收支记录参数
 */
export interface TransactionUpdateParams {
  category?: string;
  amount?: number | string;
  order_id?: number;
  note?: string;
  attachment_url?: string;
}

/**
 * 创建收支分类参数
 */
export interface TransactionCategoryCreateParams {
  name: string;
  type: 'income' | 'expense';
  description?: string;
}

/**
 * 更新收支分类参数
 */
export interface TransactionCategoryUpdateParams {
  name?: string;
  description?: string;
}

/**
 * 收支记录筛选条件
 */
export interface TransactionFilters {
  start_date?: string;
  end_date?: string;
  type?: 'income' | 'expense';
  category?: string;
  page?: number;
  page_size?: number;
}

/**
 * 收支记录列表响应
 */
export interface TransactionListResponse {
  items: Transaction[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 收支汇总统计
 */
export interface TransactionSummary {
  income: number;
  expense: number;
  net_profit: number;
  categories: Record<string, number>;
}

/**
 * 文件上传响应
 */
export interface UploadResponse {
  url: string;
  filename: string;
}

/**
 * 收支类型选项
 */
export const TRANSACTION_TYPE_OPTIONS = [
  { value: 'income', label: '收入', color: 'green' },
  { value: 'expense', label: '支出', color: 'red' },
];

/**
 * 默认收入分类
 */
export const DEFAULT_INCOME_CATEGORIES = [
  '堂食收入',
  '外卖收入',
  '自提收入',
  '其他收入',
];

/**
 * 默认支出分类
 */
export const DEFAULT_EXPENSE_CATEGORIES = [
  '食材采购',
  '房租水电',
  '员工工资',
  '设备维护',
  '其他支出',
];

/**
 * 获取收支类型显示名称
 */
export function getTransactionTypeLabel(type: string): string {
  const option = TRANSACTION_TYPE_OPTIONS.find((opt) => opt.value === type);
  return option ? option.label : type;
}

/**
 * 获取收支类型颜色
 */
export function getTransactionTypeColor(type: string): string {
  const option = TRANSACTION_TYPE_OPTIONS.find((opt) => opt.value === type);
  return option ? option.color : 'default';
}
